compaction/compaction_manager.hh

```diff
@@ -123,12 +123,19 @@ private:
     semaphore _major_compaction_sem{1};
 
     struct compaction_state {
+        // Used both by compaction tasks that refer to the compaction_state
+        // and by any function running under run_with_compaction_disabled().
+        seastar::gate gate;
```

Comment:
so we have two users of compaction_state.

compaction tasks where each has its own entry in list of tasks.

and run_with_compaction_disabled() which holds the gate.

remove(T) today will wait for the future in each task belonging to T to resolve, and also close gate for ongoing run_with_compaction_disabled() calls.

so after waiting for all tasks to complete and closing the gate, compaction_state can be finally removed from map as there's no longer any users left. from there on, compaction attempt for T should fail / be ignored, as manager was stopped on behalf of T.

with this change, we have redundant mechanism for stopping tasks for T, waiting on future and closing gate.

Reply:
That's correct.
It isn't absolutely necessary for tasks to hold the gate since they are stopped by `remove()`.
It's more of a fail safe mechanism, suspenders on top of the belt.

Reply:
I am fine with this patch too though. it can't hurt to have both, and I liked the fact that gate is held in task ctor.